wgsl/scanner.cc

```diff
@@ -0,0 +1,927 @@
+
+#include <tree_sitter/parser.h>
+#include <wctype.h>
+#include <algorithm>
+#include <bitset>
+#include <cassert>
+#include <cstring>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#if 0
+#define LOG(msg, ...) printf(msg "\n", ##__VA_ARGS__)
+#else
+#define LOG(...)
+#endif
+
+namespace {
+
+/// The possible external tokens matched by this custom scanner.
+/// The order of the entries in this enumerator must match the 'externals' in
+/// the grammar.js.
+enum Token {
+  BLOCK_COMMENT,
+  DISAMBIGUATE_TEMPLATE,  // A zero-length token used to scan ahead
+  TEMPLATE_ARGS_START,
+  TEMPLATE_ARGS_END,
+  LESS_THAN,           // '<'
+  LESS_THAN_EQUAL,     // '<='
+  SHIFT_LEFT,          // '<<'
+  SHIFT_LEFT_ASSIGN,   // '<<='
+  GREATER_THAN,        // '>'
+  GREATER_THAN_EQUAL,  // '>='
+  SHIFT_RIGHT,         // '>>'
+  SHIFT_RIGHT_ASSIGN,  // '>>='
+
+  // Special value, found found in grammar.js.
+  // https://tree-sitter.github.io/tree-sitter/creating-parsers#other-external-scanner-details
+  ERROR,
+};
+
+const char* str(Token tok) {
+  switch (tok) {
+    case Token::BLOCK_COMMENT:
+      return "BLOCK_COMMENT";
+    case Token::DISAMBIGUATE_TEMPLATE:
+      return "DISAMBIGUATE_TEMPLATE";
+    case Token::TEMPLATE_ARGS_START:
+      return "TEMPLATE_ARGS_START";
+    case Token::TEMPLATE_ARGS_END:
+      return "TEMPLATE_ARGS_END";
+    case Token::LESS_THAN:
+      return "LESS_THAN";
+    case Token::LESS_THAN_EQUAL:
+      return "LESS_THAN_EQUAL";
+    case Token::SHIFT_LEFT:
+      return "SHIFT_LEFT";
+    case Token::SHIFT_LEFT_ASSIGN:
+      return "SHIFT_LEFT_ASSIGN";
+    case Token::GREATER_THAN:
+      return "GREATER_THAN";
+    case Token::GREATER_THAN_EQUAL:
+      return "GREATER_THAN_EQUAL";
+    case Token::SHIFT_RIGHT:
+      return "SHIFT_RIGHT";
+    case Token::SHIFT_RIGHT_ASSIGN:
+      return "SHIFT_RIGHT_ASSIGN";
+    case Token::ERROR:
+      return "ERROR";
+    default:
+      return "<invalid>";
+  }
+}
+
+using CodePoint = uint32_t;
+
+static constexpr CodePoint kEOF = 0;
+
+struct CodePointRange {
+  CodePoint first;  // First code point in the interval
+  CodePoint last;   // Last code point in the interval (inclusive)
+};
+
+inline bool operator<(CodePoint code_point, CodePointRange range) {
+  return code_point < range.first;
+}
+inline bool operator<(CodePointRange range, CodePoint code_point) {
+  return range.last < code_point;
+}
+
+// Interval ranges of all code points in the Unicode 14 XID_Start set
+// This array needs to be in ascending order.
+constexpr CodePointRange kXIDStartRanges[] = {
```

Comment:
I wonder if it is possible to reuse Tree-sitter's internal existing knowledge of this code point range (used for the identifier pattern).

Reply:
Yeah, I wondered that too and mentioned it in my feedback at https://github.com/tree-sitter/tree-sitter/issues/2063.

Reply:
That'd both concise and make it easier to not make scanner part of source tree

Reply:
@ben-clayton would being able to access to `keyword_lex_fn` in TSLanguage provide a solution?

Reply:
No idea. I literally used tree sitter for the first time yesterday.

A possibly easier solution might be to find a dependency we can use that has these unicode tables. I presume tree sitter must be using some table, somewhere.